src/test/java/com/limechain/network/protocol/grandpa/GrandpaEngineTest.java

```diff
@@ -0,0 +1,238 @@
+import io.libp2p.core.PeerId;
+import io.libp2p.core.Stream;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.math.BigInteger;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import static org.mockito.ArgumentMatchers.contains;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.mockStatic;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+class GrandpaEngineTest {
+
+    private GrandpaEngine grandpaEngine;
+
+    private static final NeighbourMessage NEIGHBOUR_MESSAGE =
+            new NeighbourMessage(1, BigInteger.ONE, BigInteger.TWO, BigInteger.TEN);
+
+    byte[] encodedNeighbourMessage
+            = new byte[] {2, 1, 1, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 10, 0, 0, 0};
+
+    private Stream stream;
+    private PeerId peerId;
+    private static final SyncedState SYNCED_STATE = mock(SyncedState.class);
+    private static final ConnectionManager CONNECTION_MANAGER = mock(ConnectionManager.class);
+    private static final Logger LOGGER = mock(Logger.class);
+    private static final int ROLE = 2;
+
+    @BeforeAll
+    static void init() {
+        mockStatic(SyncedState.class).when(SyncedState::getInstance).thenReturn(SYNCED_STATE);
+        mockStatic(ConnectionManager.class).when(ConnectionManager::getInstance).thenReturn(CONNECTION_MANAGER);
+        mockStatic(Logger.class).when(() -> Logger.getLogger(GrandpaEngine.class.getName())).thenReturn(LOGGER);
+        BlockAnnounceHandshake handshake = mock(BlockAnnounceHandshake.class);
+        when(SYNCED_STATE.getHandshake()).thenReturn(handshake);
+        when(SYNCED_STATE.getNeighbourMessage()).thenReturn(NEIGHBOUR_MESSAGE);
+        when(handshake.getNodeRole()).thenReturn(ROLE);
+    }
+
+    @BeforeEach
+    void setup() {
+        grandpaEngine = new GrandpaEngine();
+        stream = mock(Stream.class);
+        peerId = mock(PeerId.class);
+        when(stream.remotePeerId()).thenReturn(peerId);
+        when(peerId.toString()).thenReturn("P1");
+    }
+
+    @Test
+    void receiveRequestWithUnknownGrandpaTypeShouldLogAndIgnore() {
+        byte[] unknownTypeMessage = new byte[] {7, 1, 1, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 10, 0, 0, 0};
+
+        grandpaEngine.receiveRequest(unknownTypeMessage, stream);
+
+        verify(LOGGER).log(eq(Level.WARNING), contains("Unknown grandpa message type"));
+    }
+
+    // INITIATOR STREAM
+    @Test
+    void receiveNonHandshakeRequestOnInitiatorStreamShouldLogAndIgnore() {
+        byte[] message = new byte[] {2, 1, 1, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 10, 0, 0, 0};
+        when(stream.isInitiator()).thenReturn(true);
+
+        grandpaEngine.receiveRequest(message, stream);
+
+        verify(LOGGER).log(eq(Level.WARNING), contains("Non handshake message on initiator grandpa stream"));
+    }
```

Comment:
These tests only test whether the warning is logged but not if the function returns.